processing/src/main/java/io/druid/query/monomorphicprocessing/SpecializationService.java

```diff
@@ -0,0 +1,391 @@
+import org.objectweb.asm.ClassReader;
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.ClassWriter;
+import org.objectweb.asm.commons.ClassRemapper;
+import org.objectweb.asm.commons.SimpleRemapper;
+import sun.misc.Unsafe;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.Field;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Manages class specialization during query processing.
+ * Usage:
+ *
+ * String runtimeShape = stringRuntimeShape.of(bufferAggregator);
+ * SpecializationState<ProcessingAlgorithm> specializationState = SpecializationService.getSpecializationState(
+ *   ProcessingAlgorithmImpl.class,
+ *   runtimeShape
+ * );
+ * ProcessingAlgorithm algorithm = specializationState.getSpecializedOrDefault(new ProcessingAlgorithmImpl());
+ * long loopIterations = new ProcessingAlgorithmImpl().run(bufferAggregator, ...);
+ * specializationState.accountLoopIterations(loopIterations);
+ *
+ * ProcessingAlgorithmImpl.class, passed as prototypeClass to {@link #getSpecializationState} methods must have public
+ * no-arg constructor and must be stateless (no fields).
+ *
+ * @see SpecializationState
+ */
+public final class SpecializationService
+{
+  private static final Logger LOG = new Logger(SpecializationService.class);
+
+  private static final Unsafe UNSAFE;
+  static {
+    try {
+      Field theUnsafe = Unsafe.class.getDeclaredField("theUnsafe");
+      theUnsafe.setAccessible(true);
+      UNSAFE = (Unsafe) theUnsafe.get(null);
+    }
+    catch (Exception e) {
+      throw new RuntimeException("Cannot access Unsafe methods", e);
+    }
+  }
+
+  /**
+   * If true, specialization is not actually done, an instance of prototypeClass is used as a "specialized" instance.
+   * Useful for analysis of generated assembly with JITWatch (https://github.com/AdoptOpenJDK/jitwatch), because
+   * JITWatch shows only classes present in the loaded JAR (prototypeClass should be), not classes generated during
+   * runtime.
+   */
+  private static final boolean fakeSpecialize = Boolean.getBoolean("fakeSpecialize");
+
+  /**
+   * Number of loop iterations, accounted via {@link SpecializationState#accountLoopIterations(long)} in
+   * {@link WindowedLoopIterationCounter} during the last hour window, after which WindowedLoopIterationCounter decides
+   * to specialize class for the specific runtimeShape. The default value is chosen to be so that the specialized
+   * class will likely be compiled with C2 HotSpot compiler with the default values of *BackEdgeThreshold options.
+   */
+  private static final long triggerSpecializationIterationsThreshold =
```

Comment:
Changed type of `triggerSpecializationIterationsThreshold` to int.

`SpecializationService` is not an established thing, so it's quite possible that when this code is battle-tested more, the design of SpecializationService will change so that `triggerSpecializationIterationsThreshold` doesn't make sense anymore or should have different semantics. So I don't want to expose it as an "official" configuration yet, because it will immediately impose some compatibility constrains.